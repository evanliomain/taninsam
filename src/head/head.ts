/**
 * @module array=>any
 */
/**
 * Return the first element of an array and return undefined
 * if the array is empty
 * @return the function to apply on the array to extract the first element
 * @example
 * ```
 * head<number>()([1, 2, 3, 4, 5]) // 1
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 3, 4, 5])
 *   .chain(head<number>())
 *   .value() // 1
 * ```
 */
export function head<T>(): (array: ReadonlyArray<T>) => T | undefined {
  return (array: ReadonlyArray<T>): T | undefined => {
    if (0 === array.length) {
      return undefined;
    }

    return array[0];
  };
}
