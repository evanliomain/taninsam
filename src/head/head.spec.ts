import { head } from './head';

describe('head function', () => {
  const input = [1, 2, 3, 4, 5];

  test('is a pure function with array', () => {
    head()(input);
    expect(input).toMatchSnapshot();
  });

  test('[1, 2, 3, 4, 5] |> head === 1', () => {
    expect(head()(input)).toMatchSnapshot();
  });

  test('[] |> head === undefined', () => {
    expect(head()([])).toMatchSnapshot();
  });
});
