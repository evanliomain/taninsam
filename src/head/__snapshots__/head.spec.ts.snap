// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`head function [] |> head === undefined 1`] = `undefined`;

exports[`head function [1, 2, 3, 4, 5] |> head === 1 1`] = `1`;

exports[`head function is a pure function with array 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;
